pyexchange/dydx.py

```diff
@@ -0,0 +1,261 @@
+                     amount=Wad(int(trade['amount'])))
+
+
+class DydxApi(PyexAPI):
+    """Dydx API interface.
+
+        Documentation available here: https://docs.dydx.exchange/#/
+
+        Startup guide here: https://medium.com/dydxderivatives/programatic-trading-on-dydx-4c74b8e86d88
+    """
+
+    logger = logging.getLogger()
+
+    def __init__(self, node: str, private_key: str):
+        assert (isinstance(node, str))
+        assert (isinstance(private_key, str))
+
+        # self.address = web3.eth.accounts.privateKeyToAccount(privateKey)['address']
+        self.client = Client(private_key=private_key, node=node)
+
+    def get_markets(self):
+        return self.client.get_pairs()['pairs']
+
+    def get_pair(self, pair: str):
+        assert (isinstance(pair, str))
+        return next(filter(lambda symbol: symbol['name'] == pair, self.get_markets()))
+
+    def _balances_to_list(self, balances) -> List:
+        balance_list = []
+
+        for token, balance in enumerate(balances):
+            if token == 0:
+                balances[str(token)]['currency'] = 'ETH'
+            elif token == 1:
+                balances[str(token)]['currency'] = 'SAI'
+            elif token == 2:
+                balances[str(token)]['currency'] = 'USDC'
+            elif token == 3:
+                balances[str(token)]['currency'] = 'DAI'
+
+            balance_list.append(balances[str(token)])
+
+        return balance_list
+
+    def get_balances(self):
+        return self._balances_to_list(self.client.get_my_balances()['balances'])
+
+    def get_orders(self, pair: str) -> List[Order]:
+        assert (isinstance(pair, str))
+
+        orders = self.client.get_my_orders(market=[pair], limit=None, startingBefore=None)
+
+        return list(map(lambda item: Order.to_order(item, pair), orders['orders']))
+
+    def deposit_funds(self, token, amount: float):
+        assert (isinstance(amount, float))
+
+        market = consts.MARKET_ETH
+
+        # determine if 6 or 18 decimals are needed for wei conversion
+        if token == 'USDC':
+            market = consts.MARKET_USDC
+        else:
+            market = consts.MARKET_ETH
+
+        tx_hash = self.client.eth.deposit(
+            market=market,
+            wei=utils.token_to_wei(amount, market)
+        )
+
+        receipt = self.client.eth.get_receipt(tx_hash)
+        return receipt
+
+    def place_order(self, pair: str, is_sell: bool, price: float, amount: float) -> str:
+        assert (isinstance(pair, str))
+        assert (isinstance(is_sell, bool))
+        assert (isinstance(price, float))
+        assert (isinstance(amount, float))
+
+        side = 'SELL' if is_sell else 'BUY'
+
+        self.logger.info(f"Placing order ({side}, amount {amount} of {pair},"
+                         f" price {price})...")
+
+        ## Need to retrieve the market_id used by a given token as all trades in DyDx use Wei as standard unit.
+        market_id = 0
```

Comment:
maaybe we should reuse pymaker's `model.Token` class and call normalize amount to get the price and make sure everything is consistent

Reply:
Looked into this, but I don't think that will work. DyDx only accepts a specific `wei` formatting, without any decimals. `Token.normal_amount` doesn't strip out the decimals or convert to `wei`. That being said, it might be worth the effort to add a method to `pymaker.model.Token` such as `convert_to_wei`